JumpscaleLibs/sal/zosv2/Zosv2.py

```diff
@@ -0,0 +1,197 @@
+        self.wg_port = wg_port
+        self.public_ip = public_ip
+
+class Zosv2(j.baseclasses.object):
+    __jslocation__ = "j.sal.zosv2"
+
+    def _init(self, **kwargs):
+        #j.clients.threebot.explorer_addr_set("explorer.testnet.grid.tf")
+        self.explorer = j.clients.threebot.explorer
+        self.config = NetworkConfig()
+
+    # consider a node up if it has received update during the last 10 minute
+    def is_up(self, node):
+        ago = j.data.time.epoch - (60 * 10)
+        return node.updated > ago
+
+    def get_node(self,ip_version):
+        #return node and public ip
+        nodes = self.explorer.actors_all.nodes.list().nodes
+        return self.find_node_public(nodes,ip_version)
+
+    def find_node_public(self, nodes, version=6):
+        # search a node that has a public ipv4 address
+        for node in filter(self.is_up, nodes):
+            for iface in node.ifaces:
+                for addr in iface.addrs:
+                    ip = ipaddress.ip_interface(addr).ip
+                    if ip.version != version:
+                        continue
+                    if ip.is_global:
+                        self.config.public_ip = str(ip)
+                        return (node, str(ip))
+        if version==4:
+            for node in filter(self.is_up, nodes):
+                ip = ipaddress.ip_interface(node.public_config.ipv4).ip
+                if ip.is_global:
+                    self.config.public_ip = str(ip)
+                    return (node, str(ip))
+            # need to check in ifaces
+    def find_free_wg_port(self, node):
+        ports = set(list(range(6000, 9000)))
+        used = set(node.wg_ports)
+        free = ports - used
+        return free.pop()
+
+    def new_reservation(self, user):
+        reservation_model = j.data.schema.get_from_url("tfgrid.workloads.reservation.1")
+        reservation = reservation_model.new()
+        reservation.customer_tid = user.id
+        reservation.data_reservation.expiration_provisioning = j.data.time.epoch + (60 * 10)  # 10 minutes
+        reservation.data_reservation.expiration_reservation = j.data.time.epoch + (60 * 10)  # 10 minutes
+        return  reservation
+
+    def get_user(self,name,email):
+        # should return error if doesn't exist
+        return self.explorer.actors_all.phonebook.get(name=name, email=email)
+
+    def check_ip_in_network(self, ip , network):
+        return netaddr.IPAddress(ip) in netaddr.IPNetwork(network)
+
+    def key_pair_get(self):
+        wg_private = public.PrivateKey.generate()
+        wg_public = wg_private.public_key
+
+        wg_private_base64 = wg_private.encode(Base64Encoder)
+        wg_public_base64 = wg_public.encode(Base64Encoder)
+
+        return wg_private_base64, wg_public_base64
+
+    def create_network(self,reservation,ip_range,node_id):
+        network = reservation.data_reservation.networks.new()
+        network.node_id = node_id
+        network.workload_id = 1
```

Comment:
You cannot know if the workload id is 1. Maybe there are plenty of other workload already define in the reservation.
Better approach would be to have a method that gives you the next free `workload_id` by walking all the workload in the reservation